ceph: Update osds in parallel

BlaineEXE opened the pull request:
Following design of https://github.com/rook/rook/blob/master/design/ceph/update-osds-in-parallel.md, update OSDs in parallel.

The max number of OSDs updated in parallel is currently fixed at 20.

**Bonus:** 10% more code coverage in pkg/operator/ceph/cluster/osd.

Manual testing:
- [x] test large cluster on AWS
  - [x] verify that multiple OSDs update in parallel
    - Note: created cluster with 80 PVCs w/ `cluster-on-pvc.yaml`, updated all of them in 53 update rounds. Cuts time by about 1/3 for updates.
  - [x] canceling reconcile by CephCluster CR works
  - [x] test upgrade
- [x] test upgrades on nodes
- [x] test updates from master -> master


**Description of your changes:**

**Which issue is resolved by this Pull Request:**
Resolves #4380

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/master/development-flow.html#commit-structure).
- [ ] **Skip Tests for Docs**: Add the flag for skipping the build if this is only a documentation change. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for the flag.
- [ ] **Skip Unrelated Tests**: Add a flag to run tests for a specific storage provider. See [test options](https://github.com/rook/rook/blob/master/INSTALL.md#test-storage-provider).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.

BlaineEXE commented on pkg/operator/ceph/cluster/osd/create.go:
```diff
@@ -0,0 +1,391 @@
+		errs.addError("error finding Kubernetes version. %v", err)
+		return util.NewSet(), nil
+	}
+	if !k8sVersion.AtLeast(version.MustParseSemantic("v1.13.0")) {
+		// TODO: discuss making this an errs.addError since users HAVE specified device sets, but
+		// the device sets will not be provisioned
+		logger.Warningf("skipping OSD on PVC provisioning. Minimum Kubernetes version required: 1.13.0. Actual version: %s", k8sVersion.String())
+		return util.NewSet(), nil
+	}
```
@leseb

BlaineEXE commented on pkg/operator/ceph/cluster/osd/create.go:
```diff
@@ -0,0 +1,391 @@
+		}
+		osdProps.storeConfig.DeviceClass = volume.CrushDeviceClass
+
+		// TODO: was this left from some personal debugging?
+		logger.Debugf("osdProps are %+v", osdProps)
```
@leseb @travisn ?

BlaineEXE commented on pkg/operator/ceph/cluster/osd/create.go:
```diff
@@ -0,0 +1,391 @@
+// Returns error only if the calling function should halt all OSD provisioning. Non-halting errors
+// are added to provisionErrors.
+func (c *Cluster) startProvisioningOverNodes(config *provisionConfig, errs *provisionErrors) (*util.Set, error) {
+	if len(c.spec.DataDirHostPath) == 0 {
+		// TODO: discuss making this an errs.addError since users HAVE specified storage config, but
+		// the device sets will not be provisioned
+		logger.Warningf("skipping osd provisioning where no dataDirHostPath is set")
+		return util.NewSet(), nil
+	}
```
@leseb @travisn?

BlaineEXE commented on pkg/operator/ceph/cluster/osd/deviceSet.go:
```diff
@@ -33,23 +33,24 @@ import (
 			continue
 		}
 		// Check if the volume claim template is specified
 		if len(deviceSet.VolumeClaimTemplates) == 0 {
+			// TODO: discuss if this should add to errs.addError
 			logger.Warningf("no volumeClaimTemplate specified for device set %q", deviceSet.Name)
 			continue
 		}
```
@leseb ?

BlaineEXE commented on pkg/operator/ceph/cluster/osd/create.go:
```diff
@@ -0,0 +1,391 @@
+		}
+
+		// Skip OSD prepare if deployment already exists for the PVC
+		// NOTE: this used to add OrchestrationStatusAlreadyCompleted for the PVC (before that
+		// it added OrchestrationStatusCompleted), but I changed the behavior to not create a
+		// status since status is really only used for new creations and not updates
+		if existingDeployments.Contains(dataSource.ClaimName) {
+			logger.Infof("skipping OSD prepare pod creation for PVC %q because OSD daemon using the PVC already exists", osdProps.crushHostname)
+			continue
+		}
```
Thoughts @travisn @leseb ?

BlaineEXE replied:
OSD update/upgrade is separate from creation, so my thinking is that prepare jobs only get created (and counted) when the jobs are running and have the possibility of actually creating new OSDs.

BlaineEXE commented on pkg/operator/ceph/cluster/osd/create.go:
```diff
@@ -0,0 +1,391 @@
+		status := OrchestrationStatus{Status: OrchestrationStatusStarting, PvcBackedOSD: true}
+		cmName := c.updateOSDStatus(osdProps.crushHostname, status)
+
+		if err := c.runPrepareJob(&osdProps, config); err != nil {
+			c.handleOrchestrationFailure(errs, osdProps.crushHostname, "%v", err)
+			// delete the status configmap now so that the failure added above won't be processed as
+			// though it is a result coming from the provisioning job
+			// NOTE: this changes prior behavior a bit. Prior behavior recorded the failure, and
+			// then set the status configmap to "Completed" so that it could be counted in the
+			// status processing without having an additional error added. Do we need to count
+			c.deleteStatusConfigMap(osdProps.crushHostname)
+			continue // do not record the status CM's name
+		}
```
Thoughts on "NOTE" @travisn @leseb?

BlaineEXE commented on pkg/operator/ceph/cluster/osd/create.go:
```diff
@@ -0,0 +1,391 @@
+	// Parsing storageClassDeviceSets and parsing it to volume sources
+	c.spec.Storage.VolumeSources = append(c.spec.Storage.VolumeSources, c.prepareStorageClassDeviceSets(errs)...)
+
+	// QUESTION: at what point are volume sources added to the CephCluster CR? I need to make sure
+	// code that does that didn't get lost in this refactor
+	c.ValidStorage.VolumeSources = c.spec.Storage.VolumeSources
```
@leseb ?

BlaineEXE replied:
What is the purpose of `c.spec.Storage.VolumeSources` to the user then? Is that to support OSDs on PVC via brownfield cases? If it doesn't have a use to the user, we might want to remove it from the spec.

BlaineEXE replied:
@jmolmo do you have any thoughts on this? `VolumeSources` seems to only be used internally and not exposed to users.

BlaineEXE commented on pkg/operator/ceph/cluster/osd/osd.go:
```diff
@@ -380,186 +243,46 @@ func (c *Cluster) getExistingOSDDeploymentsOnPVCs() (map[string]*apps.Deployment
-				continue
-			}
-		}
+	// TODO: why was there not a Progressing condition added here before?
+	// opconfig.ConditionExport(c.context, c.clusterInfo.NamespacedName(), cephv1.ConditionProgressing, v1.ConditionTrue, "ClusterProgressing", fmt.Sprintf("Processing node %s osd %d", nodeName, osd.ID))
```
@travisn @leseb ?

BlaineEXE commented on pkg/operator/ceph/cluster/osd/osd.go:
```diff
@@ -569,42 +292,50 @@ func (c *Cluster) startOSDDaemonsOnNode(nodeName string, config *provisionConfig
-	for _, osd := range osds {
-		logger.Debugf("start osd %v", osd)
-		opconfig.ConditionExport(c.context, c.clusterInfo.NamespacedName(), cephv1.ConditionProgressing, v1.ConditionTrue, "ClusterProgressing", fmt.Sprintf("Processing node %s osd %d", nodeName, osd.ID))
+	// TODO: reconcile this with Travis' changes to conditions
+	// TODO: why didn't PVC code have an equivalent condition export?
+	opconfig.ConditionExport(c.context, c.clusterInfo.NamespacedName(), cephv1.ConditionProgressing, v1.ConditionTrue, "ClusterProgressing", fmt.Sprintf("Processing node %s osd %d", nodeName, osd.ID))
```
@travisn @leseb ?

BlaineEXE commented on pkg/operator/ceph/cluster/osd/osd.go:
```diff
@@ -569,42 +292,50 @@ func (c *Cluster) startOSDDaemonsOnNode(nodeName string, config *provisionConfig
+	}
+
+	_, createErr := c.context.Clientset.AppsV1().Deployments(c.clusterInfo.Namespace).Create(context.TODO(), dp, metav1.CreateOptions{})
+	if createErr != nil {
+		if kerrors.IsAlreadyExists(createErr) {
+			if err = updateDeploymentAndWait(c.context, c.clusterInfo, dp, opconfig.OsdType, strconv.Itoa(osd.ID), c.spec.SkipUpgradeChecks, c.spec.ContinueUpgradeAfterChecksEvenIfNotHealthy); err != nil {
+				// TODO: should this register as a deployment failure? I think yes.
+				// logger.Errorf("failed to update osd deployment %d. %v", osd.ID, err)
+				return errors.Wrapf(err, "failed to update deployment for %s", osdLongName)
 			}
+			logger.Infof("updated deployment for %s", osdLongName)
 		} else {
-			logger.Infof("created deployment for osd %d", osd.ID)
+			// TODO: should this register as a deployment failure? I think yes.
+			// we failed to create job, update the orchestration status for this pvc
+			// logger.Warningf("failed to create osd deployment for node %q, osd %+v. %v", n.Name, osd, createErr)
+			return errors.Wrapf(err, "failed to create deployment for %s", osdLongName)
 		}
+	} else {
+		logger.Infof("created deployment for %s", osdLongName)
 	}
```
@leseb @travisn (two TODO questions that are basically the same)

BlaineEXE replied:
Yes. In `create.go` `createNewOSDsFromStatus` adds errors returned from here (and the parent caller) to the collection of errors. Since this is a single-purpose, low-level method, I let the calling function decide how to handle the errors.

BlaineEXE commented on pkg/operator/ceph/cluster/osd/status.go:
```diff
@@ -125,313 +151,209 @@ func parseOrchestrationStatus(data map[string]string) *OrchestrationStatus {
+			updateConfig.updateExistingOSDs(errs)
+
+		case <-minuteTicker.C:
+			// TODO: make it so that an orchestration cancel is requested on ANY CephCluster update
+			// so that we don't need to re-implement the timeout here
```
Just verifying this is still a good idea. @travisn @leseb

BlaineEXE commented on pkg/operator/ceph/cluster/osd/osd.go:
```diff
@@ -380,186 +243,46 @@ func (c *Cluster) getExistingOSDDeploymentsOnPVCs() (map[string]*apps.Deployment
 	// fully resolve the storage config and resources for this node
 	n := c.resolveNode(nodeName)
 	if n == nil {
-		logger.Errorf("node %q did not resolve to start osds", nodeName)
-		return
+		// NOTE: this now registers as a failure instead of just printing a warning message
+		return fmt.Errorf("failed to resolve node config to create/update deployment for %s", osdLongName)
 	}
```
Thoughts @travisn @leseb ?

BlaineEXE replied:
This will **not** abort the reconcile. `createNewOSDsFromStatus` calls this function and adds errors returned to the error collection. This is another low-level method that lets the calling function decide what is a critical error or not.

BlaineEXE commented on pkg/operator/ceph/cluster/osd/osd.go:
```diff
@@ -785,13 +508,59 @@ func (c *Cluster) getOSDInfo(d *apps.Deployment) ([]OSDInfo, error) {
+		script := c.Command[2]
+		varAssignment := "DEVICE=" // this variable assignment is followed by the block path
+		scanner := bufio.NewScanner(strings.NewReader(script))
+		for scanner.Scan() {
```
Yes. There is no other place where it is stored on legacy deployments.

BlaineEXE commented on pkg/daemon/ceph/client/osd.go:
```diff
@@ -305,3 +306,54 @@ func OsdListNum(context *clusterd.Context, clusterInfo *ClusterInfo) (OsdList, e
+		// that we can catch breaking changes to JSON output in CI testing. As a middle ground
+		// here, return error but also return the given OSD ID in the output in case the calling
+		// function wants to recover from this case.
+		return []int{osdID}, errors.Wrapf(err, "failed to unmarshal 'osd ok-to-stop %d' response", osdID)
```
Right. I don't either. But it is an option if we need to use it.

BlaineEXE commented on images/ceph/Makefile:
```diff
@@ -22,7 +22,7 @@ CEPH_VERSION = v15.2.9-20210224
 CEPH_VERSION = v15.2.9-20210224
 endif
-BASEIMAGE = ceph/ceph-$(GOARCH):$(CEPH_VERSION)
+BASEIMAGE = ceph/daemon-base:latest-master-devel # ceph/ceph-$(GOARCH):$(CEPH_VERSION)
```
I'm not sure. I thought we did in order to use the new `--max` flag. I guess I should test that.

BlaineEXE replied:
Wow, you're right! I'm really surprised the Ceph CLI is forward compatible like that.

BlaineEXE commented on cluster/examples/kubernetes/ceph/cluster-test.yaml:
```diff
@@ -17,6 +17,7 @@ data:
     [global]
     osd_pool_default_size = 1
     mon_warn_on_pool_no_redundancy = false
+    osd_crush_chooseleaf_type = 0
```
It forces failure domain = osd. I'm not sure we want to keep it or not. Maybe not since I can't think of super strong reasoning.

BlaineEXE commented on cluster/examples/kubernetes/ceph/csi/rbd/storageclass-test.yaml:
```diff
@@ -4,7 +4,7 @@ metadata:
   namespace: rook-ceph # namespace:cluster
 spec:
-  failureDomain: host
+  failureDomain: osd
```
I hope that it helps support the single node scenario slightly better if someone does use it in the world.

BlaineEXE commented on pkg/operator/ceph/cluster/osd/spec.go:
```diff
@@ -288,11 +292,9 @@ func (c *Cluster) makeDeployment(osdProps osdProperties, osd OSDInfo, provisionC
 	if osd.CVMode == "" {
-		osd.CVMode = "lvm"
+		return nil, errors.Errorf("failed to generate deployment for OSD %d. required CVMode is not specified for this OSD", osd.ID)
 	}
```
Return an error here since the code `getOSDInfo` sets `"lvm"` if this is empty. I ran into an issue updating a master-branch OSD that was created on node as "raw" because the `ROOK_CV_MODE` env var was unset for the new OSD. I also made changes to set this var on *all* OSD deployments now. I don't think we need to handle this with any special logic since `ROOK_CV_MODE=raw` was not backported for OSDs on nodes.

BlaineEXE replied:
The only time `CVMode` could be empty is when we are doing updates and have loaded OSDInfo from a deployment. But the `getOSDInfo` command sets `CVMode = "lvm"` if it is empty on a deployment like it always has. Otherwise, if it is a new OSD, that info is always returned from the provisioner. But if the provisioner doesn't return it due to some bug in the future, that should be an error.

leseb replied:
Ok but how can we tell if the error would be valid then? What if this is an upgrade?

BlaineEXE replied:
If this is an upgrade, `getOSDInfo()` reads the info from an existing deployment. If `CVMode` is not set on the deployment, it infers `"lvm"` and sets that on the `OSDInfo` struct returned from the function. This behavior of `getOSDInfo()` has not changed in my PR. IMO, we should not infer the OSD type in two places (both `getOSDInfo()` and here).

The change on the line here will now error in the following cases:
- the code from OSD prepare pods fails to report `CVMode` - IMO this is a good reason to error; new prepare pods should not be failing to provide this info
- the `getOSDInfo` code fails to set `CVMode` - IMO this is also a good reason to error; it means we are not properly reading info from existing deployments

The reason I changed this is because `CVMode` was not being set on updated OSD Deployments after your changes to add "raw" mode support on nodes. When a prepare pod created a new "raw" OSD, the deployment being created did not have `CVMode` populated. Because it was unpopulated, re-running the orchestration caused the "raw" OSD to be interpreted as "lvm" and the deployment update that followed resulted in a failing OSD. To fix the error I had to add `ROOK_CV_MODE` to the deployment. To prevent the error from happening again in the future, I believe it is good to error in this function if `CVMode` is unset so that we are not in the position again of ignoring a failure and causing more harm by doing so (i.e., an OSD that is updated in the wrong way and becomes broken).

leseb replied:
Ok thanks the clarification, I'm fine with returning an error in those circonstances.

BlaineEXE commented on pkg/operator/ceph/cluster/osd/update.go:
```diff
@@ -0,0 +1,345 @@
+			cephMatch = (d.Spec.Template.Spec.Containers[0].Image == desiredCephImage)
+		}
+
+		// undocumented option to aid debugging
+		if os.Getenv("ROOK_FORCE_OSD_UPDATES") == "yes-i-really-mean-it" {
+			rookMatch = false
+			cephMatch = false
+		}
```
This forces OSD deployments to update regardless of whether the rook/ceph version have changed. Is this useful to keep in for testing or if a user needs to force updates?

BlaineEXE replied:
This forces the code to update deployments even if they are already at the updated versions. I think that is separate from forcing updates for unhealthy daemons.

BlaineEXE replied:
I changed the name of the env var to be more descriptive.

leseb commented on pkg/operator/k8sutil/deployment.go:
```diff
@@ -140,10 +145,232 @@ func WaitForDeploymentToStart(clusterdContext *clusterd.Context, deployment *v1.
+func progressDeadlineExceeded(d *appsv1.Deployment) error {
+	for _, condition := range d.Status.Conditions {
+		if condition.Type == appsv1.DeploymentProgressing && condition.Reason == "ProgressDeadlineExceeded" {
+			return fmt.Errorf("gave up waiting for deployment %q to update because %q", d.Name, condition.Reason)
```
`error.Errorf()`

leseb commented on pkg/util/retry.go:
```diff
@@ -41,3 +43,46 @@ func Retry(maxRetries int, delay time.Duration, f func() error) error {
+		logger.Debugf("waiting for %s. retrying after %v", description, period)
+
+		select {
+		case <-time.After(period):
```
Should we have a cancellation context too? Or just check if the orch should be cancelled?

BlaineEXE replied:
This isn't currently called from a goroutine, so if I add a cancellation context, it won't ever be used.

leseb replied:
What about the check for cancellation then?

BlaineEXE replied:
Are you requesting to check for a cancelled reconcile from this function? I can do that, but I think that makes some of update logic bleed from the Ceph operator/controller down to the util package.

I don't anticipate this retry being more than 10 minutes in real world cases, which doesn't stall reconciliation for very long. 

The existing updatedeploymentandwait function doesn't look for cancellation. I don't think it's bad to add  a cancellation context here, but we also should add one to `updatedeploymentandwait` if we are going to do that, and then I think that's something for another PR.

TBH, I would like to clean up how the existing `UpdateDeploymentAndWait` is implemented. There are two callbacks. It seems unnecessarily complicated, and it means that some of the update logic that is unique to Ceph bleeds down into low-level helpers here. There could be a good opportunity to change all this at the same time.

leseb replied:
Sounds good for a future PR, the callbacks were made initially because other operators were using `UpdateDeploymentAndWait`.

leseb commented on pkg/util/retry.go:
```diff
@@ -41,3 +43,46 @@ func Retry(maxRetries int, delay time.Duration, f func() error) error {
+				if err != nil {
+					return errors.Wrapf(err, "timed out waiting for %s", description)
+				}
+				return fmt.Errorf("timed out waiting for %s", description)
```
`error.Errorf()`

BlaineEXE commented on pkg/daemon/ceph/client/upgrade_test.go:
```diff
@@ -297,3 +298,63 @@ func TestGetRetryConfig(t *testing.T) {
+			MockExecuteCommandWithOutputFile: func(command string, outFileArg string, args ...string) (string, error) {
+				t.Logf("command: %s %v", command, args)
+				if command != "ceph" || args[0] != "osd" {
+					panic("not a 'ceph osd' call")
```
I think it's better to have the tests panic if something happens outside of their expected parameters so that if things have to change in the future, we don't run the risk of obfuscating any errors.

BlaineEXE commented on pkg/operator/ceph/cluster/osd/create.go:
```diff
@@ -0,0 +1,413 @@
+
+	job, err := c.makeJob(*osdProps, config)
+	if err != nil {
+		return errors.Wrapf(err, "failed to generate osd provisioning job template for %s %q", nodeOrPVC, nodeOrPVCName)
```
`nodeOrPVC` is either the text `node` or `PVC`, and it's not necessary to quote that.

BlaineEXE commented on pkg/operator/ceph/cluster/osd/create.go:
```diff
@@ -0,0 +1,413 @@
+
+	if err := k8sutil.RunReplaceableJob(c.context.Clientset, job, false); err != nil {
+		if !kerrors.IsAlreadyExists(err) {
+			return errors.Wrapf(err, "failed to run provisioning job for %s %q", nodeOrPVC, nodeOrPVCName)
```
`nodeOrPVC` is either the text `node` or `PVC`, and it's not necessary to quote that.

BlaineEXE commented on pkg/operator/ceph/cluster/osd/create.go:
```diff
@@ -0,0 +1,413 @@
+		if !kerrors.IsAlreadyExists(err) {
+			return errors.Wrapf(err, "failed to run provisioning job for %s %q", nodeOrPVC, nodeOrPVCName)
+		}
+		logger.Infof("letting preexisting OSD provisioning job run to completion for %s %q", nodeOrPVC, nodeOrPVCName)
```
`nodeOrPVC` is either the text `node` or `PVC`, and it's not necessary to quote that.

BlaineEXE commented on pkg/operator/ceph/cluster/osd/create.go:
```diff
@@ -0,0 +1,413 @@
+		return nil
+	}
+
+	logger.Infof("started OSD provisioning job for %s %q", nodeOrPVC, nodeOrPVCName)
```
`nodeOrPVC` is either the text `node` or `PVC`, and it's not necessary to quote that.

BlaineEXE commented on pkg/operator/ceph/cluster/osd/osd.go:
```diff
@@ -785,13 +509,59 @@ func (c *Cluster) getOSDInfo(d *apps.Deployment) ([]OSDInfo, error) {
+			return v, nil
+		}
+	}
+	return "", fmt.Errorf("failed to find node/PVC name for OSD deployment %q: %+v", d.Name, d)
```
The thing being printed with `%+v` is a deployment, not an error. Having the field names is helpful here.

BlaineEXE commented on pkg/operator/ceph/cluster/osd/osd.go:
```diff
@@ -785,13 +509,59 @@ func (c *Cluster) getOSDInfo(d *apps.Deployment) ([]OSDInfo, error) {
+			}
+		}
+		if scanner.Err() != nil {
+			return "", errors.Wrapf(scanner.Err(), "failed to scan through activate init script for variable assignment %q", varAssignment)
```
This is a failure of the scanner. `scanner.Err()` is still nil if it reaches EOF.

BlaineEXE commented on pkg/operator/ceph/cluster/osd/status.go:
```diff
@@ -125,313 +156,215 @@ func parseOrchestrationStatus(data map[string]string) *OrchestrationStatus {
+func eventObjectName(e k8swatch.Event) string {
+	objName := "[could not determine name]"
+	objMeta, _ := meta.Accessor(e.Object)
+	if objMeta != nil {
```
We could log the error I guess, but this isn't likely to happen in the real world. Only in tests.

BlaineEXE commented on pkg/operator/ceph/cluster/osd/status.go:
```diff
@@ -125,313 +156,215 @@ func parseOrchestrationStatus(data map[string]string) *OrchestrationStatus {
+	cmClientset := c.context.Clientset.CoreV1().ConfigMaps(c.clusterInfo.Namespace)
+	cms, err := cmClientset.List(ctx, listOpts)
+	if err != nil {
+		logger.Warningf("failed to clean up any dangling OSD prepare status configmaps. failed to list OSD prepare status configmaps. %v", err)
```
I'm not sure I see a good way to delegate the cleanup like that.

Are you asking if we retry this dangling cleanup? If so, then we try the cleanup after every successful OSD provision.

BlaineEXE commented on pkg/operator/ceph/cluster/osd/update.go:
```diff
@@ -0,0 +1,345 @@
+	} else {
+		osdIDs, err = cephclient.OSDOkToStop(c.cluster.context, c.cluster.clusterInfo, osdIDQuery, maxUpdatesInParallel)
+		if err != nil {
+			logger.Debugf("OSD %d is not ok-to-stop. will try updating it again later", osdIDQuery)
```
I pretty much never mind adding more verbosity to logs. I'll change to `Infof`

BlaineEXE commented on pkg/operator/ceph/cluster/predicate.go:
```diff
@@ -132,23 +132,19 @@ func watchControllerPredicate(rookContext *clusterd.Context) predicate.Funcs {
-						logger.Info("upgrade requested, cancelling any ongoing orchestration")
-					}
+					// Set the cancellation flag to stop any ongoing orchestration
+					rookContext.RequestCancelOrchestration.Set()
```
This is related to the new design. Instead of having to keep a timeout on the OSDs, if we cancel on every CR change, then we don't have to worry about long OSD reconciles starving other updates.

leseb replied:
I still don't understand why this is necessary, maybe I've missed something in the design but I don't recall this part.

BlaineEXE replied:
We have 2 options to make sure that OSD reconciliation does not stop Rook from reconciling other resources when there are CephCluster CR updates:
1) Cancel orchestration every time the CephCluster CR is updated so that any ongoing OSD reconcile will be soon stopped. The OSD reconcile will resume on the next CR reconcile. (the option implemented here)
2) Put a timeout on the OSD reconcile that will stop it when progress hasn't been made for a while. Return an error so the OSDs can be reconciled again if there are no CR updates. If there are CR updates, now is when they will take hold.

In huddle conversations, Travis and I came to the agreement that stopping the reconcile on any CR update (not just Ceph version updates/upgrades) is preferred. User changes will take place more quickly, and we don't have to complicate the OSD design/code with a timeout.

leseb replied:
I see, it makes sense. I guess I was worried that could trigger undesired behavior if we brutally abort every and any reconcile. However, if it occurs it's probably a bug we should fix. Thanks!

BlaineEXE replied:
Agreed. As a note, I haven't seen any reconcile-storm or similar events that result in restarting the reconcile accidentally in my tests.

BlaineEXE commented on pkg/operator/ceph/cluster/osd/update.go:
```diff
@@ -0,0 +1,345 @@
+) {
+	ctx := context.TODO()
+
+	osdIDQuery, ok := c.queue.Pop()
```
I'm not following.

leseb replied:
I guess I saw you have a `doneUpdating()` method, so I was wondering why not use it instead. Also this was based on the code comment below `// no more OSDs to update` but maybe I misunderstood the purpose of `doneUpdating()`

BlaineEXE replied:
They accomplish the same thing. `Pop()` returns `ok==false` if the queue is empty.

`doneUpdating()` should return `true` when the queue is empty. I think using `doneUpdating()` would be a little easier to read, but using the `ok` result from `Pop()` is more safe if code around how `doneUpdating()` is implemented has to change in the future. I think using the `ok` result from `Pop()` is still better.

leseb replied:
Ok, but why having `doneUpdating()` then if you feel it could be clearer but at the same time not as good as  `Pop()`? It's confusing IMO.

BlaineEXE replied:
`doneUpdating()` is used in `.../osd/status.go` to check if the update is finished without calling `Pop()`. But I see your confusion, and this is a small change to make.

BlaineEXE commented on pkg/operator/ceph/cluster/osd/update.go:
```diff
@@ -0,0 +1,345 @@
+			continue
+		}
+
+		if osdID != osdIDQuery && !c.queue.Exists(osdID) {
```
The list of OSDs that are ok to stop includes the `osdIDQuery` which has been popped off the queue and therefore does not exist in the queue. But it does need to be updated. Maybe I should comment that.

BlaineEXE commented on pkg/operator/ceph/cluster/osd/update.go:
```diff
@@ -0,0 +1,345 @@
+		errs.addError("%v", errors.Wrapf(f.Error, "failed to update OSD deployment %q", f.ResourceName))
+	}
+
+	// If there were failures, don't retry them. If it's a transitory k8s/etcd issue, the next
```
I'm really not sure. The likely failures we would see are failures when k8s reports that the version of the thing we are trying to update is different and to retry updating. But I'm not sure about others.

BlaineEXE commented on pkg/operator/ceph/cluster/osd/update.go:
```diff
@@ -0,0 +1,345 @@
+		cephMatch := false
+		desiredCephImage := c.spec.CephVersion.Image
+		if len(d.Spec.Template.Spec.Containers) == 0 {
+			logger.Warningf("failed to determine if OSD Deployment %q is running the desired Ceph image (%s); assuming it needs updated. deployment has no containers specified", d.Name, desiredCephImage)
```
I would argue that it isn't because we can recover from it by updating the deployment.

BlaineEXE commented on pkg/daemon/ceph/client/osd.go:
```diff
@@ -305,3 +306,54 @@ func OsdListNum(context *clusterd.Context, clusterInfo *ClusterInfo) (OsdList, e
+	}
+
+	buf, err := NewCephCommand(context, clusterInfo, args).Run()
+	if err != nil {
```
I'm not following.

leseb replied:
I guess my point is that the command could fail for other reasons where catching `EBUSY` would be the real indication that the OSD is not ok-to-stop. However, we don't do this anywhere so I'm fine with the current approach.

BlaineEXE replied:
It seems unnecessary to me still. If the command fails for another reason other than EBUSY, what will we do other than try the command again later?

leseb replied:
We would print a different error message. But let's leave this for now.

BlaineEXE commented on pkg/operator/ceph/cluster/osd/status.go:
```diff
@@ -125,313 +156,215 @@ func parseOrchestrationStatus(data map[string]string) *OrchestrationStatus {
-		if err := controller.CheckForCancelledOrchestration(c.context); err != nil {
-			config.addError("%s", err.Error())
-			return false
+		if updateConfig.doneUpdating() && createConfig.doneCreating() {
```
Yes. But that is why we now cancel the reconcile on all CephCluster CR updates.

BlaineEXE commented on pkg/operator/ceph/cluster/osd/status.go:
```diff
@@ -125,313 +156,215 @@ func parseOrchestrationStatus(data map[string]string) *OrchestrationStatus {
+		select {
+		case event, ok := <-watcher.ResultChan():
+			if !ok {
+				logger.Infof("restarting watcher for OSD provisioning status ConfigMaps. the watcher closed the channel")
```
For users, I doubt it has much use. But since it doesn't happen much in the real world, I think it would be nice to leave it as `Info` at least until we are certain that this behaves as we need it to for the ETCD cache clears. I guess I don't think it's hurting as-is, and there is potential benefit.

leseb reviewed (approved):
I think this is mostly good to go, I've left a few questions. I'd like to clarify this https://github.com/rook/rook/pull/7386/files#r600524045 in particular. Once done we can go ahead and merge.

BlaineEXE commented:
Since Seb's last review, I changed behavior to *always* update OSD Deployments. It was an attempted optimization to only update deployments that need updating by checking the Ceph and Rook version labels on the deployments. But this would not catch cases where the CephCluster CR changed in a way that makes OSD Deployment modifications necessary. We can attempt to optimize in the future, but for now always update to make it less likely that we introduce a bug where OSDs might not get updated when they need to be. If OSD Deployments don't need updated, the banzaicloud annotation diffing we use will skip the update, making it a fast operation when reconciling clusters that don't need updates.

Also added a `continueUpgradesAfterChecksEvenIfUnhealthy` check into the code path (and a unit test for it), which was accidentally removed due to not using the preexisting `UpdateDeploymentAndWait()` fuction.
